.github/actions/verify-token/src/main.ts

```diff
@@ -0,0 +1,284 @@
+  ref: string;
+  ref_type: string;
+  actor: string;
+}
+
+async function run(): Promise<void> {
+  try {
+    /* Test locally:
+        $ env INPUT_SLSA-WORKFLOW-RECIPIENT="delegator_generic_slsa3.yml" \
+        INPUT_SLSA-UNVERIFIED-TOKEN="$(cat testdata/slsa-token)" \
+        GITHUB_EVENT_NAME="workflow_dispatch" \
+        GITHUB_RUN_ATTEMPT="1" \
+        GITHUB_RUN_ID="3789839403" \
+        GITHUB_RUN_NUMBER="198" \
+        GITHUB_WORKFLOW="delegate release project" \
+        GITHUB_SHA="8cbf4d422367d8499d5980a837cb9cc8e1e67001" \
+        GITHUB_REPOSITORY="laurentsimon/slsa-delegate-project" \
+        GITHUB_REPOSITORY_OWNER="laurentsimon" \
+        GITHUB_REF="refs/heads/main" \
+        GITHUB_REF_TYPE="branch" \
+        GITHUB_ACTOR="laurentsimon" \
+        nodejs ./dist/index.js
+    */
+
+    const workflowRecipient = core.getInput("slsa-workflow-recipient");
+    const unverifiedToken = core.getInput("slsa-unverified-token");
+
+    // Log the inputs for troubleshooting.
+    core.debug(`workflowRecipient: ${workflowRecipient}`);
+    core.debug(`unverifiedToken: ${unverifiedToken}`);
+    const parts = unverifiedToken.split(".");
+    if (parts.length !== 2) {
+      throw new Error(`malformed token: ${parts.length}`);
+    }
+
+    const bundleStr = Buffer.from(parts[0], "base64").toString("utf-8");
+    const b64Token = parts[1];
+    const bundle = JSON.parse(bundleStr);
+
+    // First, verify the signature, i.e., that it is signed by a certificate that
+    // chains up to Fulcio.
+    await sigstore.sigstore.verify(bundle, Buffer.from(b64Token));
+
+    const rawToken = Buffer.from(b64Token, "base64");
+    core.debug(`bundle: ${bundleStr}`);
+    core.debug(`token: ${rawToken}`);
+
+    interface rawTokenInterface {
+      version: number;
+      context: string;
+      builder: {
+        private_repository: boolean;
+        runner_label: string;
+        audience: string;
+      };
+      github: githubObj;
+      tool: {
+        actions: {
+          build_artifacts: {
+            path: string;
+          };
+        };
+        // NOTE: reusable workflows only support inputs of type
+        // boolean, number, or string.
+        // https://docs.github.com/en/actions/using-workflows/reusing-workflows#passing-inputs-and-secrets-to-a-reusable-workflow.
+        inputs: Map<string, Object>;
+      };
+    }
+
+    const rawTokenStr = rawToken.toString();
+    const rawTokenObj: rawTokenInterface = JSON.parse(rawTokenStr);
+
+    // Verify the version.
+    validateField("version", rawTokenObj.version, 1);
+
+    // Verify the context of the signature.
+    validateField("context", rawTokenObj.context, "SLSA delegator framework");
+
+    // Verify the intended recipient.
+    validateField(
+      "builder.audience",
+      rawTokenObj.builder.audience,
+      workflowRecipient
+    );
+
+    // Verify that the runner label is not empty.
+    validateNonEmptyField(
+      "builder.runner_label",
```

Comment:
we could also verify that it's one of the supported labels. Currently we only support `ubuntu-latest`. Shall we do a verification here?

Reply:
sgtm

Reply:
done